captum/optim/_core/objectives.py

```diff
@@ -0,0 +1,261 @@
+        self.hooks = ModuleOutputsHook(value)
+
+    def parameters(self):
+        return self.input_param.parameters()
+
+    def optimize(
+        self,
+        stop_criteria: Optional[StopCriteria] = None,
+        optimizer: Optional[optim.Optimizer] = None,
+    ):
+        r"""Optimize input based on loss function and objectives.
+        Args:
+            stop_criteria (StopCriteria, optional):  A function that is called
+                        every iteration and returns a bool that determines whether
+                        to stop the optimization.
+                        See captum.optim.typing.StopCriteria for details.
+            optimizer (Optimizer, optional):  An torch.optim.Optimizer used to
+                        optimize the input based on the loss function.
+        Returns:
+            *list* of *np.arrays* representing the **history**:
+            - **history** (*list*):
+                        A list of loss values per iteration.
+                        Length of the list corresponds to the number of iterations
+        """
+        stop_criteria = stop_criteria or n_steps(1024)
+        optimizer = optimizer or optim.Adam(self.parameters(), lr=0.025)
+        assert isinstance(optimizer, optim.Optimizer)
+
+        history = []
+        step = 0
+        while stop_criteria(step, self, history, optimizer):
+            optimizer.zero_grad()
+            loss_value = self.loss()
+            history.append(loss_value.cpu().detach().numpy())
+            (-1 * loss_value.mean()).backward()
+            optimizer.step()
+            step += 1
+
+        self.cleanup()
+        return history
+
+
+def n_steps(n: int) -> StopCriteria:
+    """StopCriteria generator that uses number of steps as a stop criteria.
+    Args:
+        n (int):  Number of steps to run optimization.
+    Returns:
+        *StopCriteria* callable
+    """
+    pbar = tqdm(total=n, unit="step")
+
+    def continue_while(step, obj, history, optim):
+        if len(history) > 0:
+            pbar.set_postfix({"Objective": f"{history[-1].mean():.1f}"}, refresh=False)
+        if step < n:
+            pbar.update()
+            return True
+        else:
+            pbar.close()
+            return False
+
+    return continue_while
+
+
+def channel_activation(target: nn.Module, channel_index: int) -> LossFunction:
+    def loss_function(targets_to_values: ModuleOutputMapping):
+        activations = targets_to_values[target]
+        assert activations is not None
+        # ensure channel_index is valid
+        assert channel_index < activations.shape[1]
+        # assume NCHW
+        # NOTE: not necessarily true e.g. for Linear layers
+        # assert len(activations.shape) == 4
+        return activations[:, channel_index, ...]
+
+    return loss_function
+
+
+def neuron_activation(
+    target: nn.Module, channel_index: int, x: int = None, y: int = None
+) -> LossFunction:
+    # ensure channel_index will be valid
```

Comment:
This and channel_activation seem also image specific
To make this more generic we can call `channel_index`, `neuron_selector` which can also be a callable function that performs 
```
activations[:, channel_index, ...]
```
for example. Can `neuron_selector` Callable argument help us to generalize `channel_activation` and  `neuron_activation ` under one umbrella ?